### PR TITLE
Enable running Python processes with worflows as inputs

### PR DESCRIPTION
**Pull request by JureZmrzlikar**

### CODE REVIEW

File: docs/CHANGELOG.rst [hunk 1]
```diff
@@ -15,6 +15,11 @@ Changed
 - Add ``username`` to ``current_user_permissions`` field of objects on API
 
 
+Fixed
+-----
+- Enable running Python processes with worflows as inputs
```
Comment (mstajdohar) [inline docs/CHANGELOG.rst]: Enable -> Support. Or just "Run Python...". Change the commit message too.

File: resolwe/process/fields.py [hunk 2]
```diff
@@ -513,9 +516,17 @@ def __init__(self, data_id, field, cache):
...
             if field_type.startswith("list:"):
-                field = ListField(ALL_FIELDS_MAP[field_type[5:]]())
+                inner_field_class = ALL_FIELDS_MAP[field_type[5:]]
+                if issubclass(inner_field_class, DataField):
+                    field = ListField(inner_field_class(data_type=field_type[10:]))
+                else:
+                    field = ListField(inner_field_class())
             else:
-                field = ALL_FIELDS_MAP[field_type]()
+                field_class = ALL_FIELDS_MAP[field_type]
```
Comment (mstajdohar) [inline resolwe/process/fields.py]: This branch is not covered in tests is it?

**Comment by JureZmrzlikar** [comment]
@mstajdohar Thanks, for your review. It made me think again and i decided to solve the problem in a bit more simplistic way (at least IMO). @jkokosar Can you please verify that this solution still works in your case?

**Comment by mstajdohar** [comment]
Much better, the code is now more readable, thanks.

File: resolwe/process/fields.py [hunk 3]
```diff
@@ -501,8 +501,11 @@ def __init__(self, data_id, field, cache):
         """
         super().__setattr__("_data_id", data_id)
 
+        if not cache:
```
Comment (dblenkus) [inline resolwe/process/fields.py]: You should set the default value of the `cache` to `None` if it is not required.

File: resolwe/process/fields.py [hunk 4]
```diff
@@ -512,7 +515,13 @@ def __init__(self, data_id, field, cache):
 
             field_type = field_descriptor["type"].rstrip(":")
 
-            if field_type.startswith("list:"):
+            if field_type.startswith("list:data"):
+                # The only case where output field is of type
```
Comment (dblenkus) [inline resolwe/process/fields.py]: Why is that?

File: resolwe/process/fields.py [hunk 5]
```diff
@@ -512,7 +515,13 @@ def __init__(self, data_id, field, cache):
 
             field_type = field_descriptor["type"].rstrip(":")
 
-            if field_type.startswith("list:"):
+            if field_type.startswith("list:data"):
```
Comment (dblenkus) [inline resolwe/process/fields.py]: Add colon at the end.

File: docs/CHANGELOG.rst [hunk 6]
```diff
@@ -15,6 +15,11 @@ Changed
 - Add ``username`` to ``current_user_permissions`` field of objects on API
 
 
+Fixed
+-----
+- Support worflows as inputs to Python processes
```
Comment (dblenkus) [inline docs/CHANGELOG.rst]: If change log entry starts with the "Support", then it is not a fix. Move it under "Added" or rephrase it to "Fix workflows to work with..."

File: resolwe/process/fields.py [hunk 7]
```diff
@@ -576,6 +585,8 @@ def __repr__(self):
 class DataField(Field):
     """Data object field."""
 
+    field_type = "data"
```
Comment (dblenkus) [inline resolwe/process/fields.py]: (This should go into a separate commit.)

File: resolwe/process/fields.py (hunk 3 above)
Comment (JureZmrzlikar) [reply resolwe/process/fields.py]: Done.

File: resolwe/process/fields.py (hunk 4 above)
Comment (JureZmrzlikar) [reply resolwe/process/fields.py]: This is just how we currently do it. There are no processes in resolwe/resolwe-bio that would have data objects as outputs (except for workflows).

File: resolwe/process/fields.py (hunk 5 above)
Comment (JureZmrzlikar) [reply resolwe/process/fields.py]: Check two lines above. :-)

File: docs/CHANGELOG.rst (hunk 6 above)
Comment (JureZmrzlikar) [reply docs/CHANGELOG.rst]: Done.

File: resolwe/process/fields.py (hunk 7 above)
Comment (JureZmrzlikar) [reply resolwe/process/fields.py]: Why? This enables the fix to work...

File: resolwe/process/fields.py (hunk 5 above)
Comment (JureZmrzlikar) [reply resolwe/process/fields.py]: > I don't like this solution, because it is not 100% correct :)

Why do you mean by "it is not 100% correct"?

File: resolwe/process/fields.py (hunk 7 above)
Comment (JureZmrzlikar) [reply resolwe/process/fields.py]: :+1:

File: resolwe/process/fields.py (hunk 5 above)
Comment (dblenkus) [reply resolwe/process/fields.py]: If someone uses type `list:dataid:`, it would match this condition instead of `elif`.

Comment (JureZmrzlikar) [reply resolwe/process/fields.py]: How about `if field_descriptor["type"].startswith("list:data:"):`?

Comment (dblenkus) [reply resolwe/process/fields.py]: Then it won't match fields of type `list:data:`, because final colon is striped in the line above.

Comment (JureZmrzlikar) [reply resolwe/process/fields.py]: It is not stripped in `field_descriptor["type"]`

Comment (dblenkus) [reply resolwe/process/fields.py]: Sorry, i haven't seen this change :) (`field_type` to `field_descriptor["type"]`).

Yes, that would also work, but you have to add a test for this or at least a comment.
